### DISCUSSION TITLE
Why is French DualFromZeroToTwo instead of DualOneOther?

### DISCUSSION THREAD

**Discussion by Fuitad** [axuno/SmartFormat#363] [General]
Heya,

I'm trying to understand why French is set to DualFromZeroToTwo instead of DualOneOther like English. I'm French Canadian and I'm having a hard time understanding why

``{0:plural:No object|{} object|{} objects} cached on file system`` is working in English but ``{0:plural:Aucun objet|{} objet|{} objets} en cache sur le système`` in French is giving me ``SmartFormat.Core.Formatting.FormattingException: Error parsing format string: Invalid number of plural parameters in PluralLocalizationFormatter``

Maybe there's something in FR-FR that explains this? But on a French Canadian point of view, this doesn't make much sense.

**Comment by axunonb** [comment]
My French is not good enough to argue about plural rules.
We're referring to the rules [published on unicode.org](https://www.unicode.org/cldr/charts/44/supplemental/language_plural_rules.html). Here you can see that English and French plural rules are in fact different. The question is, of course, whether the colloquial language adheres to the rules.

**Comment by Fuitad** [comment]
Ok I get it now. It's because the formatting can change once you get past 1 million. Sorry for the trouble!

**Comment by Fuitad** [comment]
I'm really sorry to bug you again. I've changed ``{0:plural:Aucun objet|{} objet|{} objets}`` to ``{0:plural:Aucun objet|{} objet|{} objets|{} objets}`` but I'm still getting the same invalid number of plural parameters error. Just how many parameters am I supposed to provide in French?

**Reply by axunonb** [reply]
Example:
```C#
//  new string[3] is just get array with 3 items
var result = Smart.Format(CultureInfo.GetCultureInfo("fr"), "{0:plural:une personne|deux personnes|plusieurs personnes}", new string[3]);```
// result: , "plusieurs personnes"
```

**Reply by Fuitad** [reply]
But in that example, what happens for zero?

**Reply by axunonb** [reply]
The plural word options matter. Have a look at the `DualFromZeroToTwo` delegate:
https://github.com/axuno/SmartFormat/blob/a9b0fb356274cd78ece239613a1f163e4d5b88ca/src/SmartFormat/Utilities/PluralRules.cs#L215C1-L245C6

For only 2 options we have:
```C#
var result = smart.Format(CultureInfo.GetCultureInfo("fr"), "{0:plural:{} personne|plusieurs personne}", 0);
// 0 personne
result = smart.Format(CultureInfo.GetCultureInfo("fr"), "{0:plural:{} personne|{} personnes}", 1);
// 1 personne
result = smart.Format(CultureInfo.GetCultureInfo("fr"), "{0:plural:{} personne|{} personnes}", 2);
// 2 personnes
result = smart.Format(CultureInfo.GetCultureInfo("fr"), "{0:plural:{} personne|{} personnes}", 3);
// 3 personnes
result = smart.Format(CultureInfo.GetCultureInfo("fr"), "{0:plural:{} personne|{} personnes}", 100);
// 100 personnes
```

**Comment by axunonb** [comment]
@karljj1 Please have glance at https://github.com/axuno/SmartFormat/blob/a9b0fb356274cd78ece239613a1f163e4d5b88ca/src/SmartFormat/Utilities/PluralRules.cs#L215C1-L245C6
If `n == 2` we get a failure. The switch case should read `>= 2` instead of `> 2`, shouldn't it?
Cannot believe that nobody had noticed this before...

**Reply by Fuitad** [reply]
This would explain so much about the errors I'm getting.

**Reply by axunonb** [reply]
@karljj1 Have you changed the code regarding `n == 2` in your fork?

**Reply by karljj1** [reply]
Hey. Sorry I have been on holiday for the last 2 weeks. I'll take a look after the holidays 😃

**Reply by axunonb** [reply]
Best season's greetings 🍀

**Reply by karljj1** [reply]
Looks like we are using the older version ` DualFromZeroToTwo => (n, c) => n == 0 || n == 1 ? 0 : 1;`, we have made no changes. I guess no one has encountered this issue yet. `value < 2 ? 0 : 1;` looks correct to me, am I missing something? 
French only has One and Other right?

**Reply by axunonb** [reply]
> French only has One and Other right?
Yes, I think so. Zero and one is singular, other are plural. @Fuitad Right?

The point is, that a value of 2 should be plural, but exactly 2 is not covered with the rule at all and returns an error result of -1. That doesn't make sense to me, but this has never been opened as an issue.
https://github.com/axuno/SmartFormat/blob/a9b0fb356274cd78ece239613a1f163e4d5b88ca/src/SmartFormat/Utilities/PluralRules.cs#L215C1-L245C6

**Reply by karljj1** [reply]
I see. I guess no one has ever needed 2 items in French! Your suggested change makes sense.

**Comment by Fuitad** [comment]
> > French only has One and Other right?
> > Yes, I think so. Zero and one is singular, other are plural. @Fuitad Right?

This is correct. Aucune porte (no door), une porte (one door), deux portes (two doors)
